SQL/New_patches/2021-09-22-ConsentModule.sql

```diff
@@ -0,0 +1,53 @@
+-- Create module
+INSERT INTO modules (Name, Active) VALUES ('consent', 'Y');
+
+-- Table for display information
+-- Can contain basic display info, or more complex
+-- training form
+CREATE TABLE `consent_display` (
+  `ConsentDisplayID` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `Title` text DEFAULT NULL,
+  `Media` text DEFAULT NULL,
+  `Description` text DEFAULT NULL,
+  `training` text DEFAULT NULL,
+  `Reset_period_days` int(10) unsigned DEFAULT NULL,
+  PRIMARY KEY (`ConsentDisplayID`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+-- direct_consent contains all info relevant to the eConsent
+-- form unique to each participant, but does not include the
+-- consent status given
+CREATE TABLE `direct_consent` (
+  `CandidateID` int(6) NOT NULL,
+  `ConsentGroupID` integer unsigned NOT NULL,
+  `OneTimeKey` varchar(16) DEFAULT NULL,
+  `Request_status` enum('created','sent','in_progress','complete','expired') NOT NULL DEFAULT 'created',
+  `Date_sent` date DEFAULT NULL,
+  `Data_cleared` tinyint(1) DEFAULT NULL,
+  `trainingProgress` varchar(255) DEFAULT NULL,
+  KEY `direct_consent_cand_ConsentGroupID` (`CandidateID`,`ConsentGroupID`),
+  CONSTRAINT `FK_direct_consent_1` FOREIGN KEY (`CandidateID`) REFERENCES `candidate` (`CandID`),
+  CONSTRAINT `FK_direct_consent_2` FOREIGN KEY (`ConsentGroupID`) REFERENCES `consent_group` (`ConsentGroupID`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+-- Relates different sites and consent groups to their 
+-- respective displays
+CREATE TABLE `consent_display_rel` (
+  `ConsentGroupID` int(10) unsigned NOT NULL,
+  `CenterID` int(10) unsigned DEFAULT NULL,
+  `ConsentDisplayID` int(10) unsigned NOT NULL,
+  CONSTRAINT `FK_consent_group_consent_display_rel_1` FOREIGN KEY (`ConsentGroupID`) REFERENCES `consent_group` (`ConsentGroupID`) ON DELETE CASCADE ON UPDATE CASCADE,
```

Comment:
Most rel tables have every field as a primary key. However, in this case CenterID can not be included in the primary key since it is nullable. It also cannot be (`ConsentGroupID`,`ConsentDisplayID`) because it's possible that these columns will need to hold the same values with a different CenterID. Do you have a suggestion for what the primary key should be? I think it would be good to have a constraint that ensures no duplicate rows are entered in this table but I'm having some trouble implementing it

Reply:
sounds like the PK combination needs to be (ConsentGroupID,ConsentDisplayID,CenterID). Is there a reason why CenterID is nullable?

Reply:
Null CenterID is default, i.e. it represents all sites. The CenterID would not be null if there is a site-specific change in the consent form. For example, if all sites have the same consent form except for site A, then there would be a row with a null CenterID linking to the general consent form, and a row with the CenterID for site A linking to the site-specific consent form.